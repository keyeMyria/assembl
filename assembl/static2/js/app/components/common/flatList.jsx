/* eslint-disable no-undef */
// @flow
import * as React from 'react';
import { I18n } from 'react-redux-i18n';
import throttle from 'lodash/throttle';

import Loader from './loader';
import { displayAlert } from '../../utils/utilityManager';
import { APOLLO_NETWORK_STATUS } from '../../constants';

type ItemNode = {
  node: Object
};

type FlatListProps = {
  items: {
    pageInfo: {
      endCursor: string,
      hasNextPage: boolean
    },
    edges: Array<ItemNode>
  },
  ListItem: Function | React.ComponentType<*>,
  className?: string,
  networkStatus: number,
  onEndReachedThreshold: number,
  itemData: Function,
  fetchMore: Function,
  refetch: Function,
  extractItems: Function
};

class FlatList extends React.Component<FlatListProps> {
  loading: boolean;

  static defaultProps = {
    onEndReachedThreshold: 0.8,
    itemData: () => {}
  };

  componentDidMount() {
    window.addEventListener('scroll', this.handleScroll);
    this.componentDidUpdate();
  }

  componentDidUpdate() {
    const { hash } = window.location;
    if (hash !== '') {
      const postIds = this.props.items.edges.map(edge => edge.node.id);
      const id = hash.replace('#', '').split('?')[0];
      // if post id not found in current items, load another page
      if (postIds.indexOf(id) === -1) {
        this.onEndReached();
      }
    }
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', this.handleScroll);
  }

  handleScroll = throttle(() => {
    if (!this.loading) {
      const { onEndReachedThreshold } = this.props;
      // $FlowFixMe
      const windowHeight = 'innerHeight' in window ? window.innerHeight : document.documentElement.offsetHeight;
      const body = document.body;
      const html = document.documentElement;
      // $FlowFixMe
      const docHeight = Math.max(body.scrollHeight, body.offsetHeight, html.clientHeight, html.scrollHeight, html.offsetHeight);
      const windowBottom = windowHeight + window.pageYOffset;
      const threshold = docHeight * onEndReachedThreshold;
      if (windowBottom >= threshold) {
        this.onEndReached();
      }
    }
  }, 100);

  onEndReached = throttle(() => {
    // The fetchMore method is used to load new data and add it
    // to the original query we used to populate the list
    const { fetchMore, items, extractItems, networkStatus } = this.props;
    // If no request is in flight for this query, and no errors happened. Everything is OK.
    if (networkStatus === APOLLO_NETWORK_STATUS.ready) {
      this.loading = true;
      fetchMore({
        variables: { after: items.pageInfo.endCursor || '' },
        updateQuery: (previousResult, { fetchMoreResult }) => {
          // Don't do anything if there weren't any new items
          const previousResultEntities = extractItems(previousResult);
          if (!fetchMoreResult || !previousResultEntities.pageInfo.hasNextPage) {
            return previousResult;
          }
          const fetchMoreResultEntities = extractItems(fetchMoreResult);
          fetchMoreResultEntities.edges = previousResultEntities.edges.concat(fetchMoreResultEntities.edges);
          return fetchMoreResult;
        }
      })
        .then(() => {
          this.loading = false;
        })
        .catch(() => {
          displayAlert('danger', I18n.t('error.loading'));
          this.loading = false;
        });
    }
  }, 100);

  onRefresh = () => {
    const { refetch } = this.props;
    refetch().catch(() => {
      displayAlert('danger', I18n.t('error.loading'));
    });
  };

  render() {
    const { networkStatus, items, ListItem, itemData, className } = this.props;
    if (
      items == null ||
      networkStatus === APOLLO_NETWORK_STATUS.loading ||
      networkStatus === APOLLO_NETWORK_STATUS.setVariables
    ) {
      return <Loader color="black" />;
    }
    const entities = items.edges;
    return (
      <div className={className}>
        {entities && entities.map((item, index) => <ListItem key={item.node.id || index} {...itemData(item)} node={item.node} />)}
        {networkStatus === APOLLO_NETWORK_STATUS.fetchMore && items.pageInfo.hasNextPage && <Loader color="black" />}
      </div>
    );
  }
}

export default FlatList;