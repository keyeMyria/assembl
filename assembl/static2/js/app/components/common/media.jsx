import React from 'react';
import { I18n } from 'react-redux-i18n';
import { Grid, Col, Button, Image, ResponsiveEmbed } from 'react-bootstrap';
import { EDFHacks } from '../home/video';

import { displayModal } from '../../utils/utilityManager';

class Media extends React.Component {
  static isValidDescription = description => !!(description && description !== '<p></p>');

  static Title = ({ value }) => (
    <div className="media-title-section">
      <div className="title-hyphen">&nbsp;</div>
      <h1 className="dark-title-1">{value || I18n.t('debate.survey.titleVideo')}</h1>
    </div>
  );

  static SideDescription = ({ content }) => (
    <div className="media-description">
      <div className="media-description-icon">
        <span className="assembl-icon-pepite color2">&nbsp;</span>
      </div>
      <div className="description-txt" dangerouslySetInnerHTML={{ __html: content }} />
      <div className="box-hyphen left">&nbsp;</div>
    </div>
  );

  static TopDescription = ({ content }) => (
    <div className="media-description-layer media-description-top" dangerouslySetInnerHTML={{ __html: content }} />
  );

  static ImageModal = content => () => {
    displayModal(null, content, false, null, null, true, 'large');
  };

  static Content = ({ content }) => {
    const isLocal = content[0] === '/';
    const isVideoFile = EDFHacks.srcIsVideoFile(content);
    const component = isLocal ? (
      <Image responsive src={content} />
    ) : (
      <ResponsiveEmbed a16by9>
        {isVideoFile ? (
          <video src={content} controls preload="none" /> // eslint-disable-line jsx-a11y/media-has-caption
        ) : (
          <iframe title="media" src={content} />
        )}
      </ResponsiveEmbed>
    );
    return (
      <div className="media-container">
        {isLocal ? (
          <Button onClick={Media.ImageModal(<div className="media-container">{component}</div>)}>{component}</Button>
        ) : (
          component
        )}
      </div>
    );
  };

  static BottomDescription = ({ content }) => (
    <div className="media-description-layer media-description-bottom" dangerouslySetInnerHTML={{ __html: content }} />
  );

  render() {
    const { title, descriptionTop, descriptionBottom, descriptionSide, htmlCode, noTitle } = this.props;
    const validDescriptionSide = Media.isValidDescription(descriptionSide);
    const validDescriptionTop = Media.isValidDescription(descriptionTop);
    const validDescriptionBottom = Media.isValidDescription(descriptionBottom);

    const validMedia = !!htmlCode;
    const somethingOnRight = !!(validDescriptionTop || validDescriptionBottom || validMedia || validDescriptionSide);
    const somethingOnLeft = validDescriptionSide;
    const something = !!(somethingOnLeft || somethingOnRight);
    const totalSize = 12;
    const leftSize = 4;
    const rightSize = totalSize - leftSize;
    return !noTitle || something ? (
      <section className="media-section background-light">
        <div className="max-container">
          {!noTitle && <Media.Title value={title} />}
          {something && (
            <Grid fluid>
              {somethingOnLeft && (
                <Col sm={totalSize} md={somethingOnRight ? leftSize : totalSize}>
                  {validDescriptionSide && <Media.SideDescription content={descriptionSide} />}
                </Col>
              )}
              {somethingOnRight && (
                <Col sm={totalSize} md={somethingOnLeft ? rightSize : totalSize}>
                  <div className="media-right">
                    {validDescriptionTop && <Media.TopDescription content={descriptionTop} />}
                    {validMedia && <Media.Content content={htmlCode} />}
                    {validDescriptionSide && <Media.SideDescription content={descriptionSide} />}
                    {validDescriptionBottom && <Media.BottomDescription content={descriptionBottom} />}
                  </div>
                </Col>
              )}
            </Grid>
          )}
        </div>
      </section>
    ) : null;
  }
}

export default Media;