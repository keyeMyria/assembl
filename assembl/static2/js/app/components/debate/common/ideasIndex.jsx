// @flow
import React from 'react';
import { Link } from 'react-router';
import { connect } from 'react-redux';
import classNames from 'classnames';

import { get } from '../../../utils/routeMap';
import { SECTION_INDEX_GENERATOR } from '../../../utils/section';
import { getChildrenFromArray } from '../../../utils/tree';

type ItemNode = {
  id: string,
  parentId: string,
  title: string
};

type IdeasIndexProps = {
  identifier: string,
  rootItem: string,
  items: Array<ItemNode>,
  parents: Array<number>,
  slug: string,
  className: ?string,
  indexGenerator: Function
};

export class DumbIdeasIndex extends React.Component<*, IdeasIndexProps, void> {
  static defaultProps = {
    className: '',
    indexGenerator: SECTION_INDEX_GENERATOR.alphanumericOr,
    parents: []
  };

  getTitle = (item: ItemNode, index: number) => {
    const { indexGenerator, parents } = this.props;
    const indexes = parents.slice();
    indexes.push(index);
    return `${indexGenerator(indexes)} ${item.title}`;
  };

  render() {
    const { parents, rootItem, identifier, items, slug, className } = this.props;
    const rootItems = getChildrenFromArray(rootItem, items);
    if (rootItems.length === 0) return null;
    return (
      <div className={classNames('index-container', className)}>
        {rootItems.map((item, index) => {
          const newParents = parents.slice();
          const itemIndex = index + 1;
          newParents.push(itemIndex);
          return (
            <div className="index-item-container" key={item.id}>
              <Link className="index-item" to={`${get('themeInPhase', { slug: slug, phase: identifier, themeId: item.id })}`}>
                {this.getTitle(item, itemIndex)}
              </Link>
              <DumbIdeasIndex
                parents={newParents}
                className={`sub-index sub-index-${newParents.length}`}
                items={items}
                rootItem={item.id}
                identifier={identifier}
              />
            </div>
          );
        })}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  slug: state.debate.debateData.slug
});

export default connect(mapStateToProps)(DumbIdeasIndex);