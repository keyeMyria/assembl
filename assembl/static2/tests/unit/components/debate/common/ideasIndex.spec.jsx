import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import { DumbIdeasIndex } from '../../../../../js/app/components/debate/common/ideasIndex';

describe('DumbIdeasIndex component', () => {
  it('should match a closed menu item', () => {
    const items = [
      {
        id: 'fooId',
        title: 'Foo',
        parentId: 'rootId'
      },
      {
        id: 'barId',
        title: 'Bar',
        parentId: 'fooId'
      }
    ];
    const props = {
      rootItem: 'rootId',
      items: items,
      identifier: 'survey',
      slug: 'slug'
    };
    const renderer = new ShallowRenderer();
    renderer.render(<DumbIdeasIndex {...props} />);
    const rendered = renderer.getRenderOutput();
    expect(rendered).toMatchSnapshot();
  });
});