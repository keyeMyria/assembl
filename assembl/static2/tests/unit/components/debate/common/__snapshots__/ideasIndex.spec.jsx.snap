// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DumbIdeasIndex component should match a closed menu item 1`] = `
<div
  className="index-container"
>
  <div
    className="index-item-container"
  >
    <Link
      className="index-item"
      onlyActiveOnIndex={false}
      style={Object {}}
      to="/slug/debate/survey/theme/fooId"
    >
      1. Foo
    </Link>
    <DumbIdeasIndex
      className="sub-index sub-index-1"
      identifier="survey"
      indexGenerator={[Function]}
      items={
        Array [
          Object {
            "id": "fooId",
            "parentId": "rootId",
            "title": "Foo",
          },
          Object {
            "id": "barId",
            "parentId": "fooId",
            "title": "Bar",
          },
        ]
      }
      parents={
        Array [
          1,
        ]
      }
      rootItem="fooId"
    />
  </div>
</div>
`;
